### PR TITLE
[fix] Use training type plugin hook when saving (FSDP 1/n)

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

When the accelerator saves the model, we should use the training type plugin's hook.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/accelerator.py [hunk 1]
```diff
@@ -379,7 +379,7 @@ def optimizer_state(self, optimizer: Optimizer) -> Dict[str, torch.Tensor]:
         return getattr(self.training_type_plugin, 'optimizer_state', lambda x: x.state_dict())(optimizer)
 
     def on_save(self, checkpoint: Dict[str, Union[Any, torch.Tensor]]) -> Dict[str, Union[Any, torch.Tensor]]:
```
Comment (kaushikb11) [inline pytorch_lightning/accelerators/accelerator.py:381]: @SeanNaren Would need to update the return type annotation as well.

Comment (SeanNaren) [reply pytorch_lightning/accelerators/accelerator.py:381]: Where exactly?

**Review by kaushikb11 — APPROVED** [review]

**Comment by SeanNaren** [comment]
I'll skip the changelog till all FSDP stuff is merged

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
